### ISSUE TITLE
HDF5 different input size

### ISSUE THREAD

**Issue by sunformoon** [faustomilletari/3D-Caffe#6]
Hi Fausto,

I am training the VNet on this useful 3D Caffe with HDf5 file as input.
If my input 3D data is the same size for each training sample, e.g., N * 1 * H * W * D for both data and label, where N is the batch size and H W D is the dimension of all the 3D data, this case is fine and I can train the VNet without any problems.

However, if I want to have the input with different size, let's say, the N always equals 1 and H W D are dynamic for each case, then it comes with issues for training. 
First, I checked the data HDF5 data, which has different size for different samples. Then, let's say the first training sample is 1 * 1 * H_0 * W_0 * D_0. The problem is that all the later input will be reshaped to the same size as the first training sample: 1 * 1 * H_0 * W_0 * D_0. Therefore, the loaded data for the second sample (and after) will give NaN since it may read the data out of bound. 

Could you have possible solutions for this problem?

Thank you!
Zhuotun

**Comment by faustomilletari** [comment]
Hello, 

I think caffe expects data of the same size as input, as specified in the prototxt.

I would suggest to look in the vnet repository how to feed data into caffe via python.

I can also suggest you to have a look to tensor flow.

Fausto Milletarì

**Comment by sunformoon** [comment]
Hello Fausto,

I managed to solve the original issue of different HDF5 size as input by a little bit edification of the HDF5 layer. 

However, I am stuck with another issue. I adopted the Softmaxwithloss layer instead of the Python DICE loss in order to debug with GDB on pure C/C++ code. Cause curently I couldn't manage to use GDB debugging on python and C/C++ together.

The stack looks like this after typing bt in GDB:

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff71a4478 in boost::shared_ptr<caffe::SyncedMemory>::operator!() const () at /usr/include/boost/smart_ptr/detail/operator_bool.hpp:62
62	        return px == 0;
(cuda-gdb) bt
#0  0x00007ffff71a4478 in boost::shared_ptr<caffe::SyncedMemory>::operator!() const () at /usr/include/boost/smart_ptr/detail/operator_bool.hpp:62
#1  0x00007ffff719a28f in caffe::Blob<float>::gpu_data() const () at src/caffe/blob.cpp:93
#2  0x00007ffff71ce28a in caffe::SoftmaxWithLossLayer<float>::Forward_gpu (this=0x22b3c970, bottom=..., top=...) at src/caffe/layers/softmax_loss_layer.cu:39
#3  0x00000000004167a2 in caffe::Layer<float>::Forward(std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> > const&, std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> > const&) () at ./include/caffe/layer.hpp:440
#4  0x00007ffff7166881 in caffe::Net<float>::ForwardFromTo(int, int) () at src/caffe/net.cpp:509
#5  0x00007ffff7166649 in caffe::Net<float>::ForwardPrefilled(float*) () at src/caffe/net.cpp:529
#6  0x00007ffff7166a14 in caffe::Net<float>::Forward(std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> > const&, float*) () at src/caffe/net.cpp:543
#7  0x00007ffff718bd13 in caffe::Solver<float>::Test(int) () at src/caffe/solver.cpp:306
#8  0x00007ffff718bb41 in caffe::Solver<float>::TestAll() () at src/caffe/solver.cpp:288
#9  0x00007ffff718abac in caffe::Solver<float>::Step(int) () at src/caffe/solver.cpp:192
#10 0x00007ffff718a69b in caffe::Solver<float>::Solve(char const*) () at src/caffe/solver.cpp:260
#11 0x00000000004120a8 in train() () at tools/caffe.cpp:132
#12 0x00000000004140d6 in main () at tools/caffe.cpp:310

Could you have any hints about what the problem could be?

Best,
Zhuotun

**Comment by sunformoon** [comment]
I found this problem is caused by the SoftmaxWithLoss layer. Recall that there are some warnings in the compilation. So I changed to another built-in loss for the debug, which seems Okay.

Thank you!
Zhuotun

**Comment by faustomilletari** [comment]
I would define a maximum size and then zero-pad all the volumes to be of equal size. this is actually taken care of if you use the code as it is, which relies on the data manager class.

**Comment by sunformoon** [comment]
That can be a solution. I would dig into this.
Thank you Fausto!

Zhuotun
